nselib/libssh2-utility.lua

```diff
@@ -0,0 +1,172 @@
+---
+-- Utility functions for libssh2.
+--
+-- Provides helper class for the libssh2 binding that abstracts away 
+-- details of running remote commands.
+--
+-- @copyright Same as Nmap--See https://nmap.org/book/man-legal.html
+-- @class module
+-- @name libssh2-utility
+
+local stdnse = require "stdnse"
+
+local libssh2 = stdnse.silent_require "libssh2"
+
+SSHConnection = {}
+
+---
+-- Returns a new connection object.
+--
+-- @return A connection object.
+function SSHConnection:new()
+  local o = {}
+  setmetatable(o, self)
+  self.__index = self
+  return o
+end
+
+---
+-- Sets up a connection with a server.
+--
+-- @param host A host to connect to.
+-- @param port A port to connect to.
+-- @return true on success or nil on failure
+function SSHConnection:connect(host, port)
+  self.session = libssh2.session_open(host, port.number)
+  if self.session then
+    return true
+  end
+end
+
+
+---
+-- Sets up a connection with a server.
+--
+-- @param host A host to connect to.
+-- @param port A port to connect to.
+-- @return true on success or nil on failure
+function SSHConnection:connect_pcall(host, port)
+  local status, err 
+  status, self.session, err = pcall(libssh2.session_open, host, port.number)
+  return status, err
+end
+
+---
+-- Runs a shell command on the remote host.
+--
+-- @param cmd A command to run.
+-- @return The command output.
+function SSHConnection:run_remote(cmd)
+  if not (self.session and self.authenticated) then
+    return false
+  end  
+  local channel = libssh2.open_channel(self.session)
+  libssh2.channel_exec(self.session, channel, cmd) 
+  libssh2.channel_send_eof(self.session, channel)
+  local buff = ""
+  local data = ""
+  while not libssh2.channel_eof(channel) do
+    data = libssh2.channel_read(self.session, channel)
+    if data then
+      buff = buff .. data
+    end
+  end
+  return buff
+end
+
+---
+-- Attempts to authenticate using provided credentials.
+--
+-- @param username A username to authenticate as.
+-- @param password A password to authenticate as.
+-- @return true on success or false on failure.
+function SSHConnection:password_auth(username, password)
```

Comment:
Do you want this to be added in the nse_libssh2.cc and the libssh2-utility.lua as a function or as a part of the ssh scripts as well?